core/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -424,39 +419,41 @@ public String prettyPrint() {
      *
      * @param existingAllocationId allocation id to use. If null, a fresh allocation id is generated.
      */
-    public void initialize(ShardRouting shard, String nodeId, @Nullable String existingAllocationId, long expectedSize) {
+    public ShardRouting initialize(ShardRouting shard, String nodeId, @Nullable String existingAllocationId, long expectedSize) {
         ensureMutable();
         assert shard.unassigned() : shard;
-        shard.initialize(nodeId, existingAllocationId, expectedSize);
-        node(nodeId).add(shard);
+        ShardRouting initializedShard = shard.initialize(nodeId, existingAllocationId, expectedSize);
+        node(nodeId).add(initializedShard);
         inactiveShardCount++;
-        if (shard.primary()) {
+        if (initializedShard.primary()) {
             inactivePrimaryCount++;
         }
-        addRecovery(shard);
-        assignedShardsAdd(shard);
+        addRecovery(initializedShard);
+        assignedShardsAdd(initializedShard);
+        return initializedShard;
     }
 
     /**
      * Relocate a shard to another node, adding the target initializing
      * shard as well as assigning it. And returning the target initializing
      * shard.
      */
```

Comment:
Javadocs here need to be updated to say returning both the source and target shard.